### PR TITLE
Migrate rustc_driver to SessionDiagnostic

### PR DESCRIPTION
**Pull request by adriantombu**
First timer noob here 👋🏽 I'm having a problem understanding how I can retrieve the span, and how to properly construct the error structs to avoid the current compilation errors.

Any help pointing me in the right direction would be much appreciated 🙌🏽

### CODE REVIEW

File: compiler/rustc_driver/src/lib.rs [hunk 1]
```diff
@@ -581,18 +590,24 @@ pub fn try_process_rlink(sess: &Session, compiler: &interface::Compiler) -> Comp
             sess.init_crate_types(collect_crate_types(sess, &[]));
             let outputs = compiler.build_output_filenames(sess, &[]);
             let rlink_data = fs::read(file).unwrap_or_else(|err| {
-                sess.fatal(&format!("failed to read rlink file: {}", err));
+                sess.fatal(RlinkUnableToRead {
+                    span: Default::default(),
```
Comment (davidtwco) [inline compiler/rustc_driver/src/lib.rs:594]: You don't need a `Span` if the original error didn't have one, just don't add a `span` field and don't annotate any fields with `#[primary_span]`.

File: compiler/rustc_driver/src/lib.rs [hunk 2]
```diff
@@ -1116,7 +1131,11 @@ fn extra_compiler_flags() -> Option<(Vec<String>, bool)> {
         }
     }
 
-    if !result.is_empty() { Some((result, excluded_cargo_defaults)) } else { None }
+    if !result.is_empty() {
```
Comment (davidtwco) [inline compiler/rustc_driver/src/lib.rs:1134]: Are you using `./x.py fmt`? I'm not expecting these formatting changes.

**Review by davidtwco — CHANGES_REQUESTED** [review]
This is a great start, I've left some pointers.

File: compiler/rustc_driver/src/lib.rs (hunk 2 above)
Comment (adriantombu) [reply compiler/rustc_driver/src/lib.rs:1134]: Ah my bad, I'm using `rustfmt` on file save with CLion, I will fix that!

**Comment by davidtwco** [comment]
We actually have a [no merge policy](https://rustc-dev-guide.rust-lang.org/git.html#no-merge-policy), would you be able to rebase instead?

**Comment by adriantombu** [comment]
Absolutely, sorry about that!

File: compiler/rustc_driver/src/session_diagnostics.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+use rustc_macros::SessionDiagnostic;
+
+#[derive(SessionDiagnostic)]
+#[diag(driver::rlink_unable_to_read)]
+pub(crate) struct RlinkUnableToRead {
+    pub error_message: String,
```
Comment (davidtwco) [inline compiler/rustc_driver/src/session_diagnostics.rs:6]: See #100754 for an example of just including the `io::Error` directly in these types, which I think I prefer.

File: compiler/rustc_driver/src/session_diagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+#[derive(SessionDiagnostic)]
+#[diag(driver::rlink_unable_to_deserialize)]
+pub(crate) struct RlinkUnableToDeserialize {
+    pub error_message: String,
+}
+
+#[derive(SessionDiagnostic)]
+#[diag(driver::rlink_no_a_file)]
+pub(crate) struct RlinkNotAFile {}
```
Comment (davidtwco) [inline compiler/rustc_driver/src/session_diagnostics.rs:17]:
```suggestion
pub(crate) struct RlinkNotAFile;
```

**Review by davidtwco — COMMENTED** [review]
I think this is almost ready, I've left another couple comments, thanks for your work on this

**Comment by davidtwco** [comment]
Use `@rustbot ready` when you've addressed the comments :)

File: compiler/rustc_driver/src/lib.rs [hunk 5]
```diff
@@ -581,18 +586,18 @@ pub fn try_process_rlink(sess: &Session, compiler: &interface::Compiler) -> Comp
             sess.init_crate_types(collect_crate_types(sess, &[]));
             let outputs = compiler.build_output_filenames(sess, &[]);
             let rlink_data = fs::read(file).unwrap_or_else(|err| {
-                sess.fatal(&format!("failed to read rlink file: {}", err));
+                sess.emit_fatal(RlinkUnableToRead { err });
             });
             let codegen_results = match CodegenResults::deserialize_rlink(rlink_data) {
```
Comment (davidtwco) [inline compiler/rustc_driver/src/lib.rs:591]: Do you think you could modify `deserialize_rlink` to return an enum with the different error messages so that we can translate them here?

Comment (adriantombu) [reply compiler/rustc_driver/src/lib.rs:591]: Yeah, I can surely have a try at that!

Comment (davidtwco) [reply compiler/rustc_driver/src/lib.rs:591]: I wasn't expecting that the enum returned by `deserialize_rlink` would itself derive `SessionDiagnostic` or `SessionSubdiagnostic`, but that we'd have some matching structs in `rustc_driver` for each of the error kinds and we'd use those. If we add `SessionDiagnostic` for enums later then we can tidy this up, but for now, just have `deserialize_rlink`'s error be a plain ol' enum.

Comment (adriantombu) [reply compiler/rustc_driver/src/lib.rs:591]: Haha I knew I went too far with this stupid idea of mine 🙈

File: compiler/rustc_codegen_ssa/src/lib.rs [hunk 6]
```diff
@@ -212,30 +219,34 @@ impl CodegenResults {
...
 
         let mut decoder = MemDecoder::new(&data[4..], 0);
         let rustc_version = decoder.read_str();
         let current_version = RUSTC_VERSION.unwrap();
         if rustc_version != current_version {
-            return Err(format!(
-                ".rlink file was produced by rustc version {rustc_version}, but the current version is {current_version}."
-            ));
+            return Err(CodegenErrors::RustcVersionMismatch {
+                rustc_version: rustc_version.to_string(),
```
Comment (davidtwco) [inline compiler/rustc_codegen_ssa/src/lib.rs:247]: Is it possible for us to avoid these `.to_string()` calls? It isn't a big deal because this is the error path but it would be nice.

File: compiler/rustc_macros/src/lib.rs [hunk 7]
```diff
@@ -163,6 +163,7 @@ decl_derive!(
 decl_derive!(
     [SessionSubdiagnostic, attributes(
         // struct/variant attributes
+        diag,
```
Comment (davidtwco) [inline compiler/rustc_macros/src/lib.rs:166]: Why is this necessary?

**Review by davidtwco — COMMENTED** [review]
This is great, one or two comments then we can land this.

File: compiler/rustc_macros/src/lib.rs (hunk 7 above)
Comment (adriantombu) [reply compiler/rustc_macros/src/lib.rs:166]: It's not, I forgot to remove it after my spaghetti recipe

File: compiler/rustc_codegen_ssa/src/lib.rs (hunk 6 above)
Comment (adriantombu) [reply compiler/rustc_codegen_ssa/src/lib.rs:247]: Sure thing!

Comment (adriantombu) [reply compiler/rustc_codegen_ssa/src/lib.rs:247]: I was able to remove two of them, I'm not sure if it's possible for the other two in an simple way

**Review by davidtwco — APPROVED** [review]

**Comment by adriantombu** [comment]
Thanks for your explanations and patience @davidtwco!
